File: backend/src/main/java/cn/etarch/mao/harness/mcp/McpClientManager.java
package cn.etarch.mao.harness.mcp;

import cn.etarch.mao.harness.mcp.entity.McpServer;
import cn.etarch.mao.harness.mcp.model.McpToolRef;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.modelcontextprotocol.client.McpClient;
import io.modelcontextprotocol.client.McpSyncClient;
import io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport;
import io.modelcontextprotocol.client.transport.ServerParameters;
import io.modelcontextprotocol.client.transport.StdioClientTransport;
import io.modelcontextprotocol.json.McpJsonMapper;
import io.modelcontextprotocol.json.jackson3.JacksonMcpJsonMapper;
import io.modelcontextprotocol.spec.McpSchema;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import tools.jackson.databind.json.JsonMapper;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

/**
 * CLOUD 模式 MCP 客户端会话级连接管理。
 * <p>
 * 以 {@code sessionId → serverId → McpSyncClient} 维护会话级连接：
 * - 会话构建时 {@code connectAndListTools} 建立连接并拉取工具清单；
 * - Agent 执行期间 {@code callTool} 复用连接执行工具调用；
 * - 会话结束时 {@code closeSession} 统一关闭（终止 stdio 子进程 / 释放 HTTP 连接）。
 * <p>
 * 同一台服务器在管理后台「测试连接」场景不绑定会话：连接 → 拉取清单 → 立即关闭。
 */
@Slf4j
@Component
public class McpClientManager {

    /** 会话级连接：sessionId → serverId → client */
    private final ConcurrentMap<Long, ConcurrentMap<Long, McpSyncClient>> sessionClients = new ConcurrentHashMap<>();

    private final ObjectMapper objectMapper;
    private final McpJsonMapper jsonMapper;
    private final long clientTimeoutSeconds;

    public McpClientManager(ObjectMapper objectMapper,
                            @Value("${app.mcp.client-timeout-seconds:120}") long clientTimeoutSeconds) {
        this.objectMapper = objectMapper;
        this.jsonMapper = new JacksonMcpJsonMapper(JsonMapper.builder().build());
        this.clientTimeoutSeconds = clientTimeoutSeconds;
    }

    /**
     * 建立会话级连接并拉取工具清单（CLOUD 模式）。
     * 连接失败抛出异常，由调用方降级处理（不注入该服务器工具）。
     */
    public List<McpToolRef> connectAndListTools(Long sessionId, McpServer server, Map<String, String> env) {
        McpSyncClient client = connect(server, env);
        sessionClients.computeIfAbsent(sessionId, k -> new ConcurrentHashMap<>())
                .put(server.getId(), client);
        List<McpToolRef> tools = toToolRefs(server, client.listTools());
        log.info("MCP client connected (CLOUD): session={}, server={}, tools={}", sessionId, server.getName(), tools.size());
        return tools;
    }

    /**
     * 执行 MCP 工具调用（CLOUD 模式）。
     * 连接不存在时按需懒连接（同一会话内已建立的连接会被复用）。
     *
     * @return 工具执行结果（纯文本拼接）；失败返回含 error 字段的 JSON 字符串
     */
    public String callTool(Long sessionId, Long serverId, String toolName, String argumentsJson) {
        McpSyncClient client = getClient(sessionId, serverId);
        if (client == null) {
            return "{\"error\":\"MCP connection not found for serverId=" + serverId + "\"}";
        }
        try {
            Map<String, Object> args = parseArguments(argumentsJson);
            McpSchema.CallToolResult result = client.callTool(
                    McpSchema.CallToolRequest.builder().name(toolName).arguments(args).build());
            return formatResult(result);
        } catch (Exception e) {
            log.warn("MCP callTool failed: serverId={}, tool={}, error={}", serverId, toolName, e.getMessage());
            return "{\"error\":\"MCP tool call failed: " + escapeJson(e.getMessage()) + "\"}";
        }
    }

    /**
     * 测试连接：连接 → 拉取工具清单 → 关闭连接（管理后台「测试连接」/「查看工具」）。
     *
     * @return 服务器暴露的工具列表；连接失败抛出异常，由 Controller 转为错误提示
     */
    public List<McpToolRef> testConnection(McpServer server, Map<String, String> env) {
        McpSyncClient client = connect(server, env);
        try {
            List<McpToolRef> tools = toToolRefs(server, client.listTools());
            log.info("MCP test connection OK: server={}, tools={}", server.getName(), tools.size());
            return tools;
        } finally {
            closeClient(server.getName(), client);
        }
    }

    /**
     * 关闭会话的所有 MCP 连接（会话结束 / 断连清理）。
     */
    public void closeSession(Long sessionId) {
        ConcurrentMap<Long, McpSyncClient> clients = sessionClients.remove(sessionId);
        if (clients == null) {
            return;
        }
        clients.forEach((serverId, client) -> closeClient("session-" + sessionId + "/server-" + serverId, client));
        log.info("Closed {} MCP client connections for session {}", clients.size(), sessionId);
    }

    /** 会话是否有已建立的 MCP 连接。 */
    public boolean hasSessionClients(Long sessionId) {
        ConcurrentMap<Long, McpSyncClient> clients = sessionClients.get(sessionId);
        return clients != null && !clients.isEmpty();
    }

    // ── internal ──────────────────────────────────────────────────────

    private McpSyncClient getClient(Long sessionId, Long serverId) {
        ConcurrentMap<Long, McpSyncClient> clients = sessionClients.get(sessionId);
        return clients != null ? clients.get(serverId) : null;
    }

    private McpSyncClient connect(McpServer server, Map<String, String> env) {
        io.modelcontextprotocol.spec.McpClientTransport transport = buildTransport(server, env);
        try {
            McpSyncClient client = McpClient.sync(transport)
                    .requestTimeout(Duration.ofSeconds(clientTimeoutSeconds))
                    .initializationTimeout(Duration.ofSeconds(Math.min(clientTimeoutSeconds, 60)))
                    .build();
            client.initialize();
            return client;
        } catch (Exception e) {
            throw new IllegalStateException("连接 MCP 服务器 " + server.getName() + " 失败: " + e.getMessage(), e);
        }
    }

    private io.modelcontextprotocol.spec.McpClientTransport buildTransport(McpServer server, Map<String, String> env) {
        if (McpServer.TYPE_STDIO.equals(server.getServerType())) {
            List<String> args = parseArgs(server.getArgsJson());
            ServerParameters params = ServerParameters.builder(server.getCommand())
                    .args(args)
                    .env(env)
                    .build();
            return new StdioClientTransport(params, jsonMapper);
        }
        // HTTP/SSE：优先 Streamable HTTP（同时兼容 2025-03-26 与 2025-06-18 协议）
        HttpClientStreamableHttpTransport.Builder builder = HttpClientStreamableHttpTransport.builder(server.getUrl())
                .jsonMapper(jsonMapper)
                .connectTimeout(Duration.ofSeconds(Math.min(clientTimeoutSeconds, 30)));
        return builder.build();
    }

    private List<McpToolRef> toToolRefs(McpServer server, McpSchema.ListToolsResult result) {
        List<McpToolRef> refs = new ArrayList<>();
        if (result == null || result.tools() == null) {
            return refs;
        }
        for (McpSchema.Tool tool : result.tools()) {
            refs.add(new McpToolRef(server.getId(), server.getName(),
                    tool.name(), tool.description() != null ? tool.description() : "",
                    tool.inputSchema() != null ? tool.inputSchema() : Map.of()));
        }
        return refs;
    }

    private List<String> parseArgs(String argsJson) {
        if (argsJson == null || argsJson.isBlank()) {
            return List.of();
        }
        try {
            List<String> args = objectMapper.readValue(argsJson,
                    objectMapper.getTypeFactory().constructCollectionType(List.class, String.class));
            return args != null ? args : List.of();
        } catch (Exception e) {
            log.warn("Failed to parse MCP server args, using empty: {}", e.getMessage());
            return List.of();
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseArguments(String argumentsJson) {
        if (argumentsJson == null || argumentsJson.isBlank()) {
            return Map.of();
        }
        try {
            Map<String, Object> args = objectMapper.readValue(argumentsJson, Map.class);
            return args != null ? args : Map.of();
        } catch (Exception e) {
            log.warn("Failed to parse MCP tool arguments, using empty: {}", e.getMessage());
            return Map.of();
        }
    }

    /**
     * 将 MCP 调用结果转为字符串：
     * - 服务器标记 isError → JSON 错误对象；
     * - 正常 → 拼接全部 TextContent；非文本内容（图片等）以占位描述标注；
     * - 存在 structuredContent 时优先序列化输出（结构化数据更利于 Agent 理解）。
     */
    private String formatResult(McpSchema.CallToolResult result) {
        if (result == null) {
            return "{\"error\":\"Empty MCP tool result\"}";
        }
        if (Boolean.TRUE.equals(result.isError())) {
            String text = extractText(result);
            return "{\"error\":\"" + escapeJson(text != null ? text : "MCP tool returned error") + "\"}";
        }
        if (result.structuredContent() != null) {
            try {
                return objectMapper.writeValueAsString(result.structuredContent());
            } catch (Exception ignored) {
                // fall through to text extraction
            }
        }
        StringBuilder sb = new StringBuilder();
        if (result.content() != null) {
            for (McpSchema.Content content : result.content()) {
                if (content instanceof McpSchema.TextContent textContent) {
                    if (sb.length() > 0) {
                        sb.append('\n');
                    }
                    sb.append(textContent.text());
                } else if (content instanceof McpSchema.ImageContent) {
                    if (sb.length() > 0) {
                        sb.append('\n');
                    }
                    sb.append("[图片内容：MCP 服务器返回了图片（base64），请根据上下文说明图片内容]");
                } else {
                    if (sb.length() > 0) {
                        sb.append('\n');
                    }
                    sb.append("[非文本内容：").append(content.getClass().getSimpleName()).append("]");
                }
            }
        }
        return sb.length() > 0 ? sb.toString() : "(MCP 工具返回空结果)";
    }

    private String extractText(McpSchema.CallToolResult result) {
        if (result.content() == null) {
            return null;
        }
        StringBuilder sb = new StringBuilder();
        for (McpSchema.Content content : result.content()) {
            if (content instanceof McpSchema.TextContent textContent) {
                if (sb.length() > 0) {
                    sb.append('\n');
                }
                sb.append(textContent.text());
            }
        }
        return sb.length() > 0 ? sb.toString() : null;
    }

    private void closeClient(String label, McpSyncClient client) {
        try {
            client.closeGracefully();
        } catch (Exception e) {
            log.debug("MCP client closeGracefully failed ({}), forcing close", label, e);
            try {
                client.close();
            } catch (Exception ignored) {
                // best effort
            }
        }
    }

    private static String escapeJson(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}

File: backend-ts/src/harness/mcp/mcp-client-manager.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
import { harnessLog } from '../log.js';
import { TYPE_STDIO, fullToolName, normalizeMcpInputSchema, type McpServer, type McpToolRef } from './entity/mcp-server.js';

type AnyClient = {
  connect(transport: unknown): Promise<void>;
  listTools(): Promise<{ tools?: Array<{ name: string; description?: string; inputSchema?: Record<string, unknown> }> }>;
  callTool(args: { name: string; arguments?: Record<string, unknown> }): Promise<{
    isError?: boolean;
    content?: Array<{ type?: string; text?: string }>;
    structuredContent?: unknown;
  }>;
  close(): Promise<void>;
};

export class McpClientManager {
  private readonly sessionClients = new Map<number, Map<number, AnyClient>>();

  constructor(private readonly clientTimeoutSeconds = 120) {}

  async connectAndListTools(sessionId: number, server: McpServer, env: Record<string, string>): Promise<McpToolRef[]> {
    const client = await this.connect(server, env);
    let map = this.sessionClients.get(sessionId);
    if (!map) {
      map = new Map();
      this.sessionClients.set(sessionId, map);
    }
    // 重连同一 serverId 时先关闭旧客户端，避免 STDIO 子进程泄漏
    const previous = map.get(server.id!);
    if (previous) {
      await this.closeClient(`session-${sessionId}/server-${server.id} (stale)`, previous);
    }
    map.set(server.id!, client);
    try {
      const tools = await this.toToolRefs(server, await client.listTools());
      harnessLog('info', `MCP client connected (CLOUD): session=${sessionId}, server=${server.name}, tools=${tools.length}`);
      return tools;
    } catch (e) {
      // listTools 失败时不留下悬空连接
      map.delete(server.id!);
      await this.closeClient(`session-${sessionId}/server-${server.id} (listTools failed)`, client);
      throw e;
    }
  }

  async callTool(sessionId: number | null, serverId: number, toolName: string, argumentsJson: string): Promise<string> {
    const client = sessionId != null ? this.sessionClients.get(sessionId)?.get(serverId) : undefined;
    if (!client) return JSON.stringify({ error: `MCP connection not found for serverId=${serverId}` });
    const args = parseArguments(argumentsJson);
    // 参数非法（模型输出被截断等）时兜底成 {} 会让 MCP 工具在缺参数下执行或报无关业务错，
    // 模型拿到的是误导性结果而非「参数不合法，重试」。与内置工具一致：直接返回错误、不发起调用。
    if (args == null) {
      harnessLog('warn', `MCP callTool rejected invalid arguments JSON: serverId=${serverId}, tool=${toolName}`);
      return JSON.stringify({ error: `MCP 工具参数不是合法 JSON 对象，请重新生成参数后重试: ${toolName}` });
    }
    try {
      // 必须有超时：MCP 服务器挂起会拖死整轮 Promise.all 工具执行
      const result = await this.withTimeout(
        client.callTool({ name: toolName, arguments: args }),
        this.clientTimeoutSeconds * 1000,
        `MCP tool call timed out after ${this.clientTimeoutSeconds}s: ${toolName}`,
      );
      return formatResult(result);
    } catch (e) {
      harnessLog('warn', `MCP callTool failed: serverId=${serverId}, tool=${toolName}, error=${(e as Error).message}`);
      return JSON.stringify({ error: 'MCP tool call failed: ' + escapeJson((e as Error).message) });
    }
  }

  async testConnection(server: McpServer, env: Record<string, string>): Promise<McpToolRef[]> {
    const client = await this.connect(server, env);
    try {
      const tools = await this.toToolRefs(server, await client.listTools());
      harnessLog('info', `MCP test connection OK: server=${server.name}, tools=${tools.length}`);
      return tools;
    } finally {
      await this.closeClient(server.name ?? '', client);
    }
  }

  async closeSession(sessionId: number): Promise<void> {
    const clients = this.sessionClients.get(sessionId);
    this.sessionClients.delete(sessionId);
    if (!clients) return;
    for (const [serverId, client] of clients) {
      await this.closeClient(`session-${sessionId}/server-${serverId}`, client);
    }
    harnessLog('info', `Closed ${clients.size} MCP client connections for session ${sessionId}`);
  }

  hasSessionClients(sessionId: number): boolean {
    const clients = this.sessionClients.get(sessionId);
    return clients != null && clients.size > 0;
  }

  private async connect(server: McpServer, env: Record<string, string>): Promise<AnyClient> {
    // HTTP：先 Streamable，连接失败再回退 SSE。
    // StreamableHTTPClientTransport 构造通常不抛错，协议不兼容发生在 connect 阶段，
    // 因此回退必须包在 connect 而非构造器上。
    if (server.serverType !== TYPE_STDIO) {
      const url = new URL(server.url ?? '');
      try {
        const client = new Client({ name: 'mao', version: '1.0.0' }) as unknown as AnyClient;
        await client.connect(new StreamableHTTPClientTransport(url));
        return client;
      } catch (streamableErr) {
        harnessLog('info', `MCP StreamableHTTP connect failed for ${server.name}, trying SSE: ${(streamableErr as Error).message}`);
        try {
          const client = new Client({ name: 'mao', version: '1.0.0' }) as unknown as AnyClient;
          await client.connect(new SSEClientTransport(url));
          return client;
        } catch (sseErr) {
          throw new Error(
            `连接 MCP 服务器 ${server.name} 失败: streamable=${(streamableErr as Error).message}; sse=${(sseErr as Error).message}`,
            { cause: sseErr },
          );
        }
      }
    }
    const transport = this.buildTransport(server, env);
    try {
      const client = new Client({ name: 'mao', version: '1.0.0' }) as unknown as AnyClient;
      await client.connect(transport);
      return client;
    } catch (e) {
      throw new Error(`连接 MCP 服务器 ${server.name} 失败: ${(e as Error).message}`, { cause: e });
    }
  }

  private buildTransport(server: McpServer, env: Record<string, string>): unknown {
    if (server.serverType === TYPE_STDIO) {
      const args = parseArgs(server.argsJson);
      return new StdioClientTransport({
        command: server.command ?? '',
        args,
        env: { ...process.env, ...env } as Record<string, string>,
      });
    }
    // HTTP 路径在 connect() 中处理 Streamable→SSE 回退
    return new StreamableHTTPClientTransport(new URL(server.url ?? ''));
  }

  private async toToolRefs(
    server: McpServer,
    result: { tools?: Array<{ name: string; description?: string; inputSchema?: Record<string, unknown> }> },
  ): Promise<McpToolRef[]> {
    const refs: McpToolRef[] = [];
    for (const tool of result.tools ?? []) {
      refs.push({
        serverId: server.id!,
        serverName: server.name ?? '',
        toolName: tool.name,
        description: tool.description ?? '',
        inputSchema: normalizeMcpInputSchema(tool.inputSchema),
        fullToolName: fullToolName(server.name ?? '', tool.name),
      });
    }
    return refs;
  }

  private async closeClient(label: string, client: AnyClient): Promise<void> {
    try {
      await client.close();
    } catch (e) {
      harnessLog('debug', `Failed to close MCP client ${label}: ${(e as Error).message}`);
    }
  }

  private async withTimeout<T>(promise: Promise<T>, ms: number, message: string): Promise<T> {
    let timer: ReturnType<typeof setTimeout> | undefined;
    try {
      return await Promise.race([
        promise,
        new Promise<never>((_, reject) => {
          timer = setTimeout(() => reject(new Error(message)), ms);
        }),
      ]);
    } finally {
      if (timer) clearTimeout(timer);
    }
  }
}

function parseArgs(argsJson: string | null | undefined): string[] {
  if (!argsJson || argsJson.trim() === '') return [];
  try {
    const parsed = JSON.parse(argsJson) as unknown;
    return Array.isArray(parsed) ? parsed.map(String) : [];
  } catch {
    return [];
  }
}

/** 解析工具参数；非法 JSON 或非对象返回 null（调用方据此拒绝调用），空参数视为 {}。 */
function parseArguments(argumentsJson: string | null | undefined): Record<string, unknown> | null {
  if (!argumentsJson || argumentsJson.trim() === '') return {};
  try {
    const parsed = JSON.parse(argumentsJson) as unknown;
    return parsed && typeof parsed === 'object' && !Array.isArray(parsed) ? parsed as Record<string, unknown> : null;
  } catch {
    return null;
  }
}

function formatResult(result: {
  isError?: boolean;
  content?: Array<{ type?: string; text?: string }>;
  structuredContent?: unknown;
}): string {
  if (!result) return JSON.stringify({ error: 'Empty MCP tool result' });
  if (result.isError) {
    const text = (result.content ?? []).map((c) => c.text).filter(Boolean).join('\n');
    return JSON.stringify({ error: text || 'MCP tool returned error' });
  }
  if (result.structuredContent != null) {
    try {
      return JSON.stringify(result.structuredContent);
    } catch { /* fall through */ }
  }
  const parts: string[] = [];
  for (const c of result.content ?? []) {
    if (c.type === 'text' || c.text) parts.push(c.text ?? '');
    else if (c.type === 'image') parts.push('[图片内容：MCP 服务器返回了图片（base64），请根据上下文说明图片内容]');
  }
  return parts.join('\n');
}

function escapeJson(s: string): string {
  return s.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}
